// gorsejs/dist/model/feedback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.upsertFeedbacks = exports.insertFeedbacks = exports.getFeedbacks = exports.getFeedbacksByType = exports.deleteFeedback = exports.getFeedback = void 0;
function getFeedback(axios, { type, userId, itemId, cursorOptions }) {
    return axios
        .get(`/feedback/${type}/${userId}/${itemId}`, {
        params: cursorOptions,
    })
        .then(({ data }) => {
        return data;
    });
}
exports.getFeedback = getFeedback;
function deleteFeedback(axios, { type, userId, itemId }) {
    return axios
        .delete(`/feedback/${type}/${userId}/${itemId}`)
        .then(({ data }) => {
        return data;
    });
}
exports.deleteFeedback = deleteFeedback;
function getFeedbacksByType(axios, { type, cursorOptions }) {
    return axios
        .get(`/feedback/${type}`, {
        params: cursorOptions,
    })
        .then(({ data }) => {
        return data;
    });
}
exports.getFeedbacksByType = getFeedbacksByType;
function getFeedbacks(axios, options) {
    return axios
        .get(`/feedback`, {
        params: options,
    })
        .then(({ data }) => {
        return data;
    });
}
exports.getFeedbacks = getFeedbacks;
function insertFeedbacks(axios, feedbacksList) {
    return axios
        .post(`/feedback`, feedbacksList)
        .then(({ data }) => {
        return data.RowAffected;
    });
}
exports.insertFeedbacks = insertFeedbacks;
function upsertFeedbacks(axios, feedbacksList) {
    return axios
        .put(`/feedback`, feedbacksList)
        .then(({ data }) => {
        return data.RowAffected;
    });
}
exports.upsertFeedbacks = upsertFeedbacks;
// TODO - Item Feedbacks
// TODO - User Feedbacks
